Update CI image with develop docker image

lsy323 opened the pull request:
Update CI image with the develop docker image.

Plan to take the following steps to update CI image:
- [x] Enable PyTorch/XLA build in GHA workflows with new docker image.
- [x] Enable PyTorch/XLA build & tests in GHA workflows with new docker image.
- [x] Create PR in upstream to update the CI image used in upstream CI for PyTorch/XLA tests: https://github.com/pytorch/pytorch/pull/109757

cc @JackCaoG

JackCaoG commented on .github/workflows/_build.yml:
```diff
@@ -80,7 +80,7 @@ jobs:
         run: |
           # pid=$(docker run -t -d -w "$WORKDIR" "${GCR_DOCKER_IMAGE}")
-          pid=$(docker run --gpus all --privileged -t -d -w "$WORKDIR" -e "TERM=xterm-256color" "${GCR_DOCKER_IMAGE}")
+          pid=$(docker run --privileged -t -d -w "$WORKDIR" -e "TERM=xterm-256color" "${GCR_DOCKER_IMAGE}")
```
I think CI is running on a GPU machine.. at least we run the test on gpu machine

lsy323 replied:
The build machine is CPU I think. Only GPU tests are run on GPU machine. I saw this in the log `runner: linux.12xlarge`

PyTorch is built with USE_CUDA=0 in a CUDA container and PyTorch/XLA is built with XLA_CUDA=1. See https://github.com/pytorch/xla/blob/44033edb7b3f4b6a94497ad6083508c5f2dbb838/.circleci/docker/Dockerfile#L15

JackCaoG replied:
ah sg!

lsy323 commented:
Build and CPU tests passed with CUDA dev container.

However, the GPU test `TestPjRtDistributedDataParallel.test_ddp_correctness_env_init` in `test/pjrt/test_ddp.py` is failing. It seems that it turned into a bad state, causing the other test got stuck.

Try skipping ddp test to see if any other test is failing in GPU dev container.

yeounoh commented on .circleci/build.sh:
```diff
@@ -11,22 +10,6 @@ clone_pytorch $PYTORCH_DIR $XLA_DIR
 # Use bazel cache
 USE_CACHE=1
 
-SCCACHE="$(which sccache)"
```
Let's leave this as-is, this is used in the upstream (nice to have the option to turn it on and off).

yeounoh replied:
Check this out, and leave this declaration as well https://github.com/pytorch/xla/blob/ea5073b4c445f14b1a6ecdfa86101b0d7da68a0d/.github/workflows/_build.yml#L89C45-L89C59

JackCaoG replied:
hmm, I thought upstream also migrated to use our bazel remote cache, we even share the service key with them for this purpose.

lsy323 replied:
Will preserve sccache as it might be used in upstream and we may re-enable it in the future

yeounoh replied:
@JackCaoG not for pytorch build, i think. It's a few lines of code, so might as well leave as an option?

yeounoh commented on .circleci/build.sh:
```diff
@@ -41,14 +24,15 @@ apply_patches
+export USE_CUDA=0
 python setup.py install
 
-sccache --show-stats
```
also leave this.

yeounoh commented on .github/workflows/_build.yml:
```diff
@@ -64,30 +64,29 @@ jobs:
-            # push to the upstream.
-            docker tag "${GCR_DOCKER_IMAGE}" "${ECR_DOCKER_IMAGE_BASE}:v1.0" >/dev/null
-            docker push "${ECR_DOCKER_IMAGE_BASE}:v1.0" >/dev/null
+      # - name: Stage image to ECR
```
TODO (@lsy323 ): re-stage, after testing is done.

yeounoh commented on .github/workflows/_build.yml:
```diff
@@ -64,30 +64,29 @@ jobs:
       - name: Prepare build env
         shell: bash
         run: |
-          echo "declare -x SCCACHE_BUCKET=${SCCACHE_BUCKET}" | docker exec -i "${pid}" sh -c "cat >> env"
```
Don't remove `SCCACHE_BUCKET`

yeounoh commented on .github/workflows/_test.yml:
```diff
@@ -63,7 +63,7 @@ jobs:
     env:
       DOCKER_IMAGE: ${{ inputs.docker-image }}
-      WORKDIR: /var/lib/jenkins/workspace
+      WORKDIR: /home
```
the home dir is set to `/root`, so something like `/root/workspace` ? Likewise, you are picking up the htmlcov from `/root/htmlcov`.

lsy323 replied:
makes sense, updated to `/root/workspace`

yeounoh commented on .circleci/test.sh:
```diff
@@ -26,5 +26,5 @@ function install_torchvision() {
 
 export GCLOUD_SERVICE_KEY_FILE="$XLA_DIR/default_credentials.json"
-export SILO_NAME='cache-silo-ci-gcc-11'  # cache bucket for CI
+export SILO_NAME='cache-silo-ci-dev-container-python38-bullseye'  # cache bucket for CI
```
nit. Let's just use the image tag as prefix to `cache-silo-ci-dev-`, so it would look like `cache-silo-ci-dev-3.8_cuda_12.1`.

lsy323 replied:
sg, let me update the cache name

yeounoh commented on .kokoro/Dockerfile:
```diff
@@ -47,7 +47,7 @@ ARG SCCACHE="$(which sccache)"
 WORKDIR /pytorch/xla
 ARG GCLOUD_SERVICE_KEY_FILE="/pytorch/xla/default_credentials.json"
-ARG SILO_NAME='cache-silo-ci-gcc-11'  # cache bucket for CI
+ARG SILO_NAME='cache-silo-ci-dev-container-python38-bullseye'  # cache bucket for CI
```
https://github.com/pytorch/xla/pull/5290/files#r1339193970

lsy323 replied:
sg, will update the name here

yeounoh commented on .github/workflows/build_and_test.yml:
```diff
@@ -77,7 +77,7 @@ jobs:
   push-docs:
     name: "Build & publish docs"
-    if: github.event_name == 'push' && (github.event.ref == 'refs/heads/master' || startsWith(github.event.ref, 'refs/tags/r'))
+    # if: github.event_name == 'push' && (github.event.ref == 'refs/heads/master' || startsWith(github.event.ref, 'refs/tags/r'))
```
Thanks for testing, TODO(@lsy323 ) remove/re-enable after testing.

yeounoh commented on .github/workflows/build_and_test.yml:
```diff
@@ -43,14 +43,14 @@ jobs:
   test-cpu-coverage:
     name: "Collect CPU test coverage"
-    if: github.event_name == 'push' && github.event.ref == 'refs/heads/master'
+    # if: github.event_name == 'push' && github.event.ref == 'refs/heads/master'
```
Btw, make sure you rebase the master, i pushed fix for coverage test yesterday.

lsy323 replied:
Thanks for the reminder, yes I think this PR includes your fix to the coverage test

yeounoh reviewed (commented):
Left some comments, will approve after testing is complete.

lsy323 commented:
Synced with @yeounoh offline, since we have additional dependencies for CI (e.g. sccache), which are not required by development. So we create an other CI image based on the dev image, installing the necessary dependencies in the CI image. The CI Dockerfile will reuse the `.circleci/docker/Dockerfile`.

I updated the `.circleci/docker/Dockerfile` to use dev container as base image, now the CI is passing on CPU and GPU.

I had to make the following changes to Dockerfile to make it work
- Remove the `jenkins` user, otherwise I'm seeing permission error when executing the `./install_valgrind.sh`. Tried to move jenkins creation at the begining but somehow didn't help.
- Removed conda installation, it's not needed anymore

lsy323 commented:
Also we cannot create jenkins user at the end of the Dockerfile, otherwise, we need `sudo` to install pytorch, pytorch/xla

lsy323 commented on .circleci/docker/Dockerfile:
```diff
@@ -1,13 +1,13 @@
 # This requires cuda & cudnn packages pre-installed in the base image.
```
Hey @yeounoh Can you please take a pass at the updated Dockerfile? I built a new one it can pass the tests now.

lsy323 commented on .github/workflows/build_and_test.yml:
```diff
@@ -19,8 +19,7 @@ jobs:
       ecr-docker-image-base: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/xla_base
-      gcr-docker-image: gcr.io/tpu-pytorch/xla_base:latest
-      disable_xrt: 1
+      gcr-docker-image: gcr.io/tpu-pytorch/xla_base:dev-3.8_cuda_12.1
```
@yeounoh Ultimately, do we want to push the new image to `latest` tag? Or just use the current one for CI usage?

yeounoh commented on .circleci/docker/Dockerfile:
```diff
@@ -37,38 +37,13 @@ ENV CXX "${cxx}"
+COPY ./install_valgrind.sh install_valgrind.sh
 RUN bash ./install_valgrind.sh
 
-# Sets up jenkins user.
```
We use `jenkins` user for the upstream, we can add the jenkins as a sudo user for the dev docker, instead of deleting it. @lsy323

lsy323 replied:
Thanks for the reminder, yes, let's keep jenkins user

yeounoh commented on .circleci/docker/Dockerfile:
```diff
@@ -37,38 +37,13 @@ ENV CXX "${cxx}"
-    chown jenkins /home/jenkins
-RUN echo 'jenkins ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
-
-RUN mkdir -p /opt/conda /opt/cargo /opt/rustup /workspace /var/lib/jenkins && \
```
These paths are also used in the upstream template, if we can reuse, less customization for them.

lsy323 replied:
Sure, let's keep them

lsy323 commented:
After offline discussion with @yeounoh, the following changes are made since last checkpoint:

1. Preserve the `jenkins` user(As a normal user instead of root), as it's used in upstream CI template
2. Because of change 1, we need to add `--user` to `python setup.py` command, this wasn't needed before, since we were using conda and conda env is private for the `jenkins` user.
3. In Dockerfile, change exec permission for bazel and install `sudo` package

lsy323 commented:
The upstream CI failed when compiling custom_op test (which we don't compile in our CI). The root cause of the compilation failure is:

In PyTorch CI [script](https://github.com/pytorch/pytorch/blob/f1b94461aabec129164685f230d914e789ac42ba/.ci/pytorch/build.sh#L271), it passes the system python site package folder(`/usr/local/lib/python3.8/site-packages`) path to the `cmake` cmd, but the torch is installed to the user specific python package path e.g. `/home/jenkins/.python/site_packages`.

To fix this, we have 2 choices:
1. Make `jenkins` a root user
2. Keep using conda in our CI image

lsy323 commented:
After switching to conda, I hit a clang problem in which the conda has some clang dependencies installed by default, but the version is mismatched with the lib installed in the system.

To resolve this problem, I added system lib path to `LD_LIBRARY_PATH` and give them higher priority than the conda lib path. So we try to use the compilation tools installed the in the system instead of conda. `cmake` and `ninja` are removed from conda installation script because they are installed in the system.

vanbasten23 commented on .circleci/common.sh:
```diff
@@ -123,7 +102,7 @@ function install_deps_pytorch_xla() {
   XLA_DIR=$1
   pushd "$XLA_DIR"
-  python setup.py install
+  python setup.py install --user
```
I wonder why we start to install in the users' directory now

lsy323 replied:
I found `--user` is needed when jenkins is a normal user and python env is system python. Since we keep using conda now, the `--user` isn't needed anymore because the python env is setup by jenkins

vanbasten23 commented on .circleci/common.sh:
```diff
@@ -92,27 +92,6 @@ function install_deps_pytorch_xla() {
 
   sudo ln -s "$(command -v bazelisk)" /usr/bin/bazel
 
-  # Install gcc-11
```
I wonder why this section is removed

lsy323 replied:
Previously, install gcc-11 from setup script. With this update, we are using gcc-10 that is standard in debian:bullseye, so we don't need to install gcc again in our setup script

vanbasten23 commented on .circleci/docker/install_conda.sh:
```diff
@@ -41,9 +41,6 @@ function install_and_setup_conda() {
   /usr/bin/yes | pip install --upgrade numba
   /usr/bin/yes | pip install cloud-tpu-client
   /usr/bin/yes | pip install expecttest==0.1.3
-  /usr/bin/yes | pip install ninja  # Install ninja to speedup the build
```
I wonder why we stop using ninja

lsy323 replied:
Because I found it's already in base image

vanbasten23 commented on .circleci/docker/install_valgrind.sh:
```diff
@@ -9,7 +9,7 @@ tar -xjf valgrind-${VALGRIND_VERSION}.tar.bz2
 cd valgrind-${VALGRIND_VERSION}
 ./configure --prefix=/usr/local
 make -j6
-sudo make install
```
I assume the reason why we can drop "sudo" is because we run the docker with "--privileged"?

lsy323 replied:
I found the script is executed by `root` in dockerfile, which is already in sudo

vanbasten23 commented on .circleci/docker/Dockerfile:
```diff
@@ -37,38 +37,15 @@ ENV CXX "${cxx}"
 # Whether to build for TPUVM mode
 ENV TPUVM_MODE "${tpuvm}"
 
-# Rotate nvidia repo public key (last updated: 04/27/2022)
```
I wonder why we remove this part (line40-line52)

lsy323 replied:
I don't think any CUDA related packages need to be installed anymore, as base image should already have everything to build ptxla with cuda support

vanbasten23 replied:
ah that makes sense. Thanks!

vanbasten23 commented on .github/workflows/_build.yml:
```diff
@@ -96,8 +95,7 @@ jobs:
       - name: Build
         shell: bash
         run: |
-          docker exec -u jenkins "${pid}" bash -c ". ~/.bashrc && .circleci/build.sh"
```
"~/.bashrc" can be removed because it has been done in the dockerfile?

lsy323 replied:
correct

vanbasten23 commented on .github/workflows/_build.yml:
```diff
@@ -73,12 +73,12 @@ jobs:
-            docker tag "${GCR_DOCKER_IMAGE}" "${ECR_DOCKER_IMAGE_BASE}:v1.0" >/dev/null
-            docker push "${ECR_DOCKER_IMAGE_BASE}:v1.0" >/dev/null
+            docker tag "${GCR_DOCKER_IMAGE}" "${ECR_DOCKER_IMAGE_BASE}:v1.1-lite" >/dev/null
+            docker push "${ECR_DOCKER_IMAGE_BASE}:v1.1-lite" >/dev/null
```
Do you know why we add `-lite` as opposed to increment to for example v1.1?

lsy323 replied:
Hmm, because our base image is just python:debian + cuda, which is much more light weight than nvidia docker images. It's just naming haha

yeounoh reviewed (approved):
LGTM

vanbasten23 commented:
This is awesome. Thanks Siyuan!
